pegasus/helpers/section_api_helpers.rb

```diff
@@ -526,38 +526,73 @@ def self.fields
   def self.assign_script_to_section(script_id, section_id)
-    # create userscripts for users that don't have one yet
-    Dashboard.db[:user_scripts].
-      insert_ignore.
-      import(
-        [:user_id, :script_id],
-        Dashboard.db[:followers].
-          select(:student_user_id, script_id.to_s).
-          where(section_id: section_id, deleted_at: nil)
-      )
+    student_user_ids = Dashboard.db[:followers].
+      select(:student_user_id).
+      where(section_id: section_id, deleted_at: nil).
+      map {|f| f[:student_user_id]}
+    DashboardUserScript.assign_script_to_users(script_id, student_user_ids)
   end
 
+  # Assigns a script to the user via user_scripts, creating a new user_scripts object if necessary.
+  # The method noops if a user_scripts already exists with assigned_at set.
+  # @param script_id [Integer] The dashboard ID of the script.
+  # @param user_id [Integer] The dashboard ID of the user.
   def self.assign_script_to_user(script_id, user_id)
-    # creates a userscript for a user if they don't have it yet
-    Dashboard.db[:user_scripts].
-      insert_ignore.
-      import(
-        [:user_id, :script_id],
-        Dashboard.db[:users].
-          select(user_id, script_id.to_s).
-          where(id: user_id, deleted_at: nil)
+    time_now = Time.now
+    existing = Dashboard.db[:user_scripts].where(user_id: user_id, script_id: script_id).first
+    if existing
+      return if existing[:assigned_at]
+      Dashboard.db[:user_scripts].where(user_id: user_id, script_id: script_id).update(
+        updated_at: time_now,
+        assigned_at: time_now
+      )
+    else
+      Dashboard.db[:user_scripts].insert(
+        user_id: user_id,
+        script_id: script_id,
+        created_at: time_now,
+        updated_at: time_now,
+        assigned_at: time_now
       )
+    end
   end
 
+  # Assigns a script to a set of users via user_scripts, creating new user_scripts objects if
+  # necessary. The method noops for those user_scripts that already exist with assigned_at set.
+  # WARNING: This method does not verify that the users exist (aren't deleted).
   def self.assign_script_to_users(script_id, user_ids)
+    # NOTE: This method could be more simply written by iterating over user_ids, calling
+    # DashboardUserScript#assign_script_to_user for each. This (more complex) approach is used for
+    # its better DB performance.
     return if user_ids.empty?
-    # create userscripts for users that don't have one yet
+
+    time_now = Time.now
+    all_existing = Dashboard.db[:user_scripts].where(user_id: user_ids, script_id: script_id)
+    all_existing_user_ids = all_existing.map {|user_script| user_script[:user_id]}
+
+    missing_assigned_at = []
+    all_existing.each do |existing|
+      missing_assigned_at << existing[:id] unless existing[:assigned_at]
+    end
+    Dashboard.db[:user_scripts].where(id: missing_assigned_at).update(
```

Comment:
This is actually necessary independent of the bug this PR fixes, so I'm not sure a comment is necessary. In particular, a user could start work on a script before being assigned the script.